trio/_core/_io_windows.py

```diff
@@ -73,22 +82,138 @@
 # - other completion keys are available for user use
 
 
+# The completion keys we use
+class CKeys(enum.IntEnum):
+    AFD_POLL = 0
+    WAIT_OVERLAPPED = 1
+    LATE_CANCEL = 2
+    USER_DEFINED = 3  # and above
+
+
 def _check(success):
     if not success:
         raise_winerror()
     return success
 
 
+def _get_base_socket(sock):
+    if hasattr(sock, "fileno"):
+        sock = sock.fileno()
+    base_ptr = ffi.new("HANDLE *")
+    out_size = ffi.new("DWORD *")
+    failed = ws2_32.WSAIoctl(
+        ffi.cast("SOCKET", sock),
+        WSAIoctls.SIO_BASE_HANDLE,
+        ffi.NULL,
+        0,
+        base_ptr,
+        ffi.sizeof("HANDLE"),
+        out_size,
+        ffi.NULL,
+        ffi.NULL,
+    )
+    if failed:
+        code = ws2_32.WSAGetLastError()
+        raise_winerror(code)
+    return base_ptr[0]
+
+
+def _afd_helper_handle():
+    # The "AFD" driver is exposed at the NT path "\Device\Afd". We're using
+    # the Win32 CreateFile, though, so we have to pass a Win32 path. \\.\ is
+    # how Win32 refers to the NT \GLOBAL??\ directory, and GLOBALROOT is a
+    # symlink inside that directory that points to the root of the NT path
+    # system. So by sticking that in front of the NT path, we get a Win32
+    # path. Alternatively, we could use NtCreateFile directly, since it takes
+    # an NT path. But we already wrap CreateFileW so this was easier.
+    # References:
+    #   https://blogs.msdn.microsoft.com/jeremykuhne/2016/05/02/dos-to-nt-a-paths-journey/
+    #   https://stackoverflow.com/a/21704022
+    rawname = r"\\.\GLOBALROOT\Device\Afd\Trio".encode("utf-16le") + b"\0\0"
+    rawname_buf = ffi.from_buffer(rawname)
+
+    handle = kernel32.CreateFileW(
+        ffi.cast("LPCWSTR", rawname_buf),
+        FileFlags.SYNCHRONIZE,
+        FileFlags.FILE_SHARE_READ | FileFlags.FILE_SHARE_WRITE,
+        ffi.NULL,  # no security attributes
+        FileFlags.OPEN_EXISTING,
+        FileFlags.FILE_FLAG_OVERLAPPED,
+        ffi.NULL,  # no template file
+    )
+    if handle == INVALID_HANDLE_VALUE:  # pragma: no cover
+        raise_winerror()
+    return handle
+
+
+# AFD_POLL has a finer-grained set of events than other APIs. We collapse them
+# down into Unix-style "readable" and "writable".
+#
+# There's also a AFD_POLL_LOCAL_CLOSE that we could wait on, to potentially
+# catch some cases where someone forgot to call notify_closing. But it's not
+# reliable – e.g. if the socket has been dup'ed, then closing one of the
+# handles doesn't trigger the event – and it's not available on Unix-like
+# platforms. So it seems like including it here would be more likely to mask
+# subtle bugs than to actually help anything.
+
+READABLE_FLAGS = (
+    AFDPollFlags.AFD_POLL_RECEIVE
```

Comment:
Should `RECEIVE_EXPEDITED` be in here too? (I'm guessing that's for TCP urgent data)

Reply:
Right, `RECEIVE_EXPEDITED` is for urgent/OOB data. And we definitely don't want to treat the socket as readable every time we receive urgent/OOB data: https://sandstorm.io/news/2015-04-08-osx-security-bug